### ISSUE TITLE
Python 3 migration

### ISSUE THREAD

**Issue by alastair** [MTG/freesound#1083]
> As a developer and maintainer of Freesound, I want Freesound codebase to run on Python 3 before the Python 2 end of life

Opening a ticket to put out some ideas about the process that we have to take in order to get freesound running on python3

* Check that all of our dependencies run on python3
* Slowly convert our code to be able to run on both python2 and python3
  * Use `__future__` imports to convert syntax and behaviour (print, division, absolute imports)
  * Use six for changed types/modules
* Revise data boundaries for bytes/string data types, use of `unicode()`, `.encode()`, `.decode()`, etc
* Once the code is python3 syntax compatible, start running tests on python3 to see if there are any failures
* Ensure that we have tests for as many pages and processing functions as possible
* At some stage, switch to running python3!

**Comment by alastair** [comment]
Here's a list of current dependencies that don't install on python3:

 * BeautifulSoup (only used once in messages, we should replace with bs4)
 * python-memcached (upgrade available)
 * gearman (gearman3 should be a drop in replacement)
 * akismet (upgrade available)
 * pyyaml (doesn't compile on 3.7, upgrade available)
 * python-cjson (doesn't compile on 3.7, no upgrade available, used for solr, we should replace with newer client)

**Comment by Bomme** [comment]
I propose to use to follow [this guide](https://docs.python.org/3/howto/pyporting.html) in upgrading.
Specifically, I think it would be best to use [futurize](https://python-future.org/automatic_conversion.html) to get a codebase that supports both Python 2.7 and Python 3.3+.
We can upgrade with futurize in two stages to limit the number of changes to review.

[Stage 1](https://python-future.org/futurize.html#stage-1-safe-fixes)

> This applies fixes that modernize Python 2 code without changing the effect of the code. With luck, this will not introduce any bugs into the code, or will at least be trivial to fix. The changes are those that bring the Python code up-to-date without breaking Py2 compatibility. 

[Stage 2](https://python-future.org/futurize.html#stage-2-py3-style-code-with-wrappers-for-py2): 

> The goal for stage 2 is to make further mostly safe changes to the Python 2 code to use Python 3-style code that then still runs on Python 2 with the help of the appropriate builtins and utilities in `future`.

**Comment by Bomme** [comment]
I checked again which dependencies need to be updated for Py3 support, similar to @alastair (see above)

- [ ] `akismet==0.2.0`  # Can be upgraded to [1.0.1](https://akismet.readthedocs.io/en/1.0.1/upgrade.html)
- [ ] `ConcurrentLogHandler==0.9.1`  # Seems to be superseded by [`concurrent-log-handler`](https://pypi.org/project/concurrent-log-handler/)
- [ ] `feedparser==5.1.3`  # can be installed with setuptools<=58; should be upgraded soon
- [ ] `psycopg2==2.7.3.2`  # does not install, unsure why, could be my setup. The next release I could install was [2.7.5](https://www.psycopg.org/articles/tag/news/)
- [ ] `python-cjson==1.1.0`  # does not install, will most likely [not be available for Py3](https://github.com/AGProjects/python-cjson/issues/6). Can be removed after #1614
- [ ] `python-memcached==1.53`  # Can probably upgrade to the [latest available version](https://github.com/linsomniac/python-memcached/releases)

Additionally, using [`caniusepython3`](https://github.com/brettcannon/caniusepython3) I found a few more suspiciously outdated packages: 

- [ ]   `django-oauth2-provider==0.2.6.1`  # seems to be not used, since we use django-oauth-toolkit instead
- [ ]   `django-piston`  # is this used? I can not find anything 
- [ ]   `recaptcha-client`  # after looking at the source code, I doubt that it will run in Py3 even if it installs with pip3. [django-recaptcha](https://github.com/torchbox/django-recaptcha) seems to be an alternative

**Comment by ffont** [comment]
Thanks @Bomme for the work!

Re `django-oauth2-provider` I think this can be safely removed, we stopped using it after some upgrade of DRF many years ago
Re `django-piston` this was used for the first version of the api, which reached end of life many years ago as well. Can be removed.
Re `recaptcha-client` I guess our best option is to change package. We could also take this opportunity to stop using Google's recaptcha and start using hCaptcha or something like that, which seems to be more privacy friendly, but still "proprietary" so maybe not really an improvement.

**Comment by Bomme** [comment]
We run Python 3 now! #1661
